### ISSUE TITLE
bug: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B640F8 is being destroyed

### ISSUE THREAD

**Issue by ipanocloud** [apache/apisix#9698]
### Current Behavior

When conducting a load test with 200 concurrent requests, APIsix is experiencing a high number of errors.

### Expected Behavior

APIsix is not generating any error logs

### Error Logs
2023/06/19 17:39:46 [error] 43#43: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B640F8 is being destroyed
 2023/06/19 17:39:13 [error] 43#43: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B64578 is being destroyed
2023/06/19 17:38:42 [error] 43#43: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B640F8 is being destroyed
2023/06/19 17:33:25 [error] 43#43: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B640F8 is being destroyed
2023/06/19 17:24:56 [error] 43#43: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B64F08 is being destroyed
2023/06/19 17:24:54 [error] 43#43: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B64F08 is being destroyed
2023/06/19 16:59:27 [error] 43#43: *4407572 [lua] client.lua:155: ping(): Agent ping fails, failed to parse domain: failed to query the DNS server: dns server error: 3 name error, context: ngx.timer
2023/06/19 16:59:27 [error] 43#43: *4407572 [lua] resolver.lua:47: parse_domain(): failed to parse domain: skywalking.skywalking.svc.cluster.local, error: failed to query the DNS server: dns server error: 3 name error, context: ngx.timer
2023/06/19 16:51:18 [error] 43#43: *4380029 [lua] client.lua:155: ping(): Agent ping fails, failed to parse domain: failed to query the DNS server: dns server error: 3 name error, context: ngx.timer
2023/06/19 16:51:18 [error] 43#43: *4380029 [lua] resolver.lua:47: parse_domain(): failed to parse domain: skywalking.skywalking.svc.cluster.local, error: failed to query the DNS server: dns server error: 3 name error, context: ngx.timer
2023/06/19 16:49:46 [error] 44#44: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B67308 is being destroyed
2023/06/19 16:45:26 [error] 43#43: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B64578 is being destroyed
2023/06/19 16:44:31 [error] 44#44: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B671E8 is being destroyed
2023/06/19 16:38:04 [error] 44#44: in lua semaphore gc wait queue is not empty while the semaphore 00007F2C17B67308 is being destroyed

### Steps to Reproduce

1. run apisix&etcd image in k8s
2. a load test with 200 concurrent requests

### Environment

- APISIX version (run `apisix version`): 2.15.0
- Operating system (run `uname -a`): Linux apisix-7dc879cc4-hwm7g 3.10.0-1160.15.2.el7.x86_64
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`): openresty/1.21.4.1
- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`):  3.5.0
- APISIX Dashboard version, if relevant:
- Plugin runner version, for issues related to plugin runners: no
- LuaRocks version, for installation issues (run `luarocks --version`):

**Comment by leslie-tsang** [comment]
@Revolyssup Please take a look

**Comment by Revolyssup** [comment]
This error is triggered here[1] when lua gc tries to run the finalizer function to cleanup the semaphore object but the semaphore object doesn't have an empty queue. APISIX will signal the GC to delete the semaphore here[2].  
Also can you confirm if you see the same errors with use_grpc[3] set to true.

1. https://github.com/openresty/lua-nginx-module/blob/f5c2d36101ec4c64400c240dc867987f33c96346/src/ngx_http_lua_semaphore.c#L566
2. https://github.com/apache/apisix/blob/fc5ae822d26451cd9241425f14fd4ca22de8af02/apisix/core/config_etcd.lua#L172
3. https://github.com/apache/apisix/blob/fc5ae822d26451cd9241425f14fd4ca22de8af02/conf/config-default.yaml#L640

**Comment by Revolyssup** [comment]
@ipanocloud Also can you give a little details about what are the exact requests used in the load test?

**Comment by Revolyssup** [comment]
Okay I have narrowed it down to a race condition.
Each post() call here[1] posts the event to be handled inside wait() here[2]. For usual traffic, all wait() exit before this[3] line could be executed. But for large number of concurrent requests which might be triggering large number of concurrent requests to etcd, this[3] line executes before all wait() calls could be completed. Which means some of the semaphores still have something in their queue and lua triggers garbage collection of all semaphores causing the errors.

1. https://github.com/apache/apisix/blob/fc5ae822d26451cd9241425f14fd4ca22de8af02/apisix/core/config_etcd.lua#L170C9-L170C20
2. https://github.com/apache/apisix/blob/fc5ae822d26451cd9241425f14fd4ca22de8af02/apisix/core/config_etcd.lua#L280
3. https://github.com/apache/apisix/blob/fc5ae822d26451cd9241425f14fd4ca22de8af02/apisix/core/config_etcd.lua#L172

**Comment by Revolyssup** [comment]
@ipanocloud Can you confirm if you see the same issue in version 3.x.x ?

**Comment by ipanocloud** [comment]
> 

only 2.15.0, use_grpc: false  but 2.15.0 not support etcd grpc

**Comment by ipanocloud** [comment]
> @ipanocloud Can you confirm if you see the same issue in version 3.x.x ?

only 2.15.0, use_grpc: false but 2.15.0 not support etcd grpc

**Comment by ipanocloud** [comment]
> Okay I have narrowed it down to a race condition. Each post() call here[1] posts the event to be handled inside wait() here[2]. For usual traffic, all wait() exit before this[3] line could be executed. But for large number of concurrent requests which might be triggering large number of concurrent requests to etcd, this[3] line executes before all wait() calls could be completed. Which means some of the semaphores still have something in their queue and lua triggers garbage collection of all semaphores causing the errors.
> 
> Instead of setting wait_init to nil, we can set each semaphore for each key nil individually here[4], after we are sure that wait() is completed
> 
> 1. https://github.com/apache/apisix/blob/fc5ae822d26451cd9241425f14fd4ca22de8af02/apisix/core/config_etcd.lua#L170C9-L170C20
> 2. https://github.com/apache/apisix/blob/fc5ae822d26451cd9241425f14fd4ca22de8af02/apisix/core/config_etcd.lua#L280
> 3. https://github.com/apache/apisix/blob/fc5ae822d26451cd9241425f14fd4ca22de8af02/apisix/core/config_etcd.lua#L172
> 4. https://github.com/apache/apisix/blob/fc5ae822d26451cd9241425f14fd4ca22de8af02/apisix/core/config_etcd.lua#L282

I have added a custom directory "/apps" to etcd and used listening. I'm not sure if they are related.

**Comment by ipanocloud** [comment]
> @ipanocloud Also can you give a little details about what are the exact requests used in the load test?
I have added a custom directory "/apps" to etcd and used listening. I'm not sure if they are related
`function _M.init()
    local err
    apps, err = core.config.new("/apps", {
        automatic = true,
        filter = function(app)
            core.log.debug("filter app: ", core.json.delay_encode(app, true))
        end,
    })
    if not apps then
        core.log.error("failed to sync /apps: ", err)
    end
end`

**Comment by Revolyssup** [comment]
@ipanocloud Where and on which table have you added this init() function?

**Comment by kingluo** [comment]
@Revolyssup 

> But for large number of concurrent requests which might be triggering large number of concurrent requests to etcd, this[3] line executes before all wait() calls could be completed. 

I am a bit confused about your race condition guess. The number of watches is determined by the number of resource types, not the QPS. The watching is independent of the request handling path.

Anyways, I am still checking the issue.

**Comment by Revolyssup** [comment]
> @Revolyssup
> 
> > But for large number of concurrent requests which might be triggering large number of concurrent requests to etcd, this[3] line executes before all wait() calls could be completed.
> 
> I am a bit confused about your race condition guess. The number of watches is determined by the number of resource types, not the QPS. The watching is independent of the request handling path.
> 
> Anyways, I am still checking the issue.

Even if there is no correlation between number of concurrent requests in this test scenario and number of resources being watched, there seemed to be the race condition unless I have understood the flow between .post() and .wait() incorrectly. But since this is on release/2.15.0, that whole hypothesis might be wrong.

**Comment by ipanocloud** [comment]
> @ipanocloud Where and on which table have you added this init() function?

A custom plugin, priority: 94.
function _M.init()

**Comment by ipanocloud** [comment]
> @ipanocloud Where and on which table have you added this init() function?

apisix/constants.lua file, new added HTTP_ETCD_DIRECTORY={ ..., ["/apps"] = true}

**Comment by kingluo** [comment]
@Revolyssup IMO, no race condition in the new etcd watching.
I check the code and the above logs, and the issue should come from lua-resty-dns-cleint lib:
https://github.com/api7/lua-resty-dns-client/blob/master/src/resty/dns/client.lua#L25
Could you help to have a look there and try to check if this issue is a bug or not? Maybe you should try to reproduce this issue in your env. Thanks.

**Comment by kingluo** [comment]
@ipanocloud Hi, do you happen to enable etcd auth? Just to confirm and exclude the issue of etcd stuff.

**Comment by ipanocloud** [comment]
> @ipanocloud Hi, do you happen to enable etcd auth? Just to confirm and exclude the issue of etcd stuff.

no, use

**Comment by ipanocloud** [comment]
> @ipanocloud Hi, do you happen to enable etcd auth? Just to confirm and exclude the issue of etcd stuff.

no,
etcd:
  host:
    - "http://apisix-etcd.apisix-prod.svc.cluster.local:2379"
  prefix: /apisix

**Comment by Revolyssup** [comment]
@ipanocloud Can you tell me the repro steps? Which endpoint did you hit while load testing? What were your load testing parameters?

**Comment by Revolyssup** [comment]
@ipanocloud And can you confirm if the same issue is with APISIX master or APISIX 3.*.*

**Comment by Revolyssup** [comment]
> snc-app-info

Is this all the code for the plugin you added?

**Comment by Revolyssup** [comment]
@ipanocloud Can you also provide the route configuration that you used? Because I am unable to repro it even on 1000 concurrent requests.

**Comment by ipanocloud** [comment]
> @ipanocloud And can you confirm if the same issue is with APISIX master or APISIX 3.* ?

I not sure, we only use v2.15.0

**Comment by ipanocloud** [comment]
> @ipanocloud Can you also provide the route configuration that you used? Because I am unable to repro it even on 1000 concurrent requests.

Can you provide email? github show route configuration not Secure.

**Comment by Revolyssup** [comment]
> > @ipanocloud Can you also provide the route configuration that you used? Because I am unable to repro it even on 1000 concurrent requests.
> 
> Can you provide email? github show route configuration not Secure.

You can give a similar route configuration, not exact for reproducability.

**Comment by Sn0rt** [comment]
@leslie-tsang pls assign the label with can't reproduce

**Comment by ipanocloud** [comment]
> > > @ipanocloud Can you also provide the route configuration that you used? Because I am unable to repro it even on 1000 concurrent requests.
> > 
> > 
> > Can you provide email? github show route configuration not Secure.
> 
> You can give a similar route configuration, not exact for reproducability.

``
"plugins": {
    "api-restriction": {
      "allow_degradation": false,
      "disable": false,
      "header_schema": {
        "appId": "x-app-id"
      },
      "rejected_code": 401,
      "rejected_msg": "{\"code\":\"00030\",\"msg\":\"error\",\"debug\":{\"traceId\":\"N/A\",\"bizName\":\"API Gateway\",\"timestamp\":\"${timestamp}\"}}"
    }
}
``

**Comment by Sn0rt** [comment]
no reproduce@2.15

# retry this in APISIX 2.15

remove the before version

```
helm install apisix apisix/apisix --version "0.11.2"
```

verify the version

```
$ docker exec -it kind-control-plane bash                                                                                                  [09:09:35]
root@kind-control-plane:/# crictl exec -it 44e2f6cbd57c5 bash
bash-5.1# apisix version
/usr/local/openresty/luajit/bin/luajit ./apisix/cli/apisix.lua version
2.15.0
```

update `configmap`

```
nginx_config:    # config for render the template to genarate nginx.conf
      error_log: test_error.log
      error_log_level: "debug"    # warn,error
```

restart the pod

```
$ kubectl delete pod apisix-fbc449d4-vggsg                                                                                                 [09:21:20]
pod "apisix-fbc449d4-vggsg" deleted

$ kubectl get pod                                                                                                                          [09:22:10]
NAME                    READY   STATUS    RESTARTS      AGE
apisix-etcd-0           1/1     Running   2 (84m ago)   85m
apisix-etcd-1           1/1     Running   0             85m
apisix-etcd-2           1/1     Running   2 (84m ago)   85m
apisix-fbc449d4-bs8tz   1/1     Running   0             44s
```

```bash
$ export NODE_PORT=$(kubectl get --namespace default -o jsonpath="{.spec.ports[0].nodePort}" services apisix-gateway)
  export NODE_IP=$(kubectl get nodes --namespace default -o jsonpath="{.items[0].status.addresses[0].address}")
  echo http://$NODE_IP:$NODE_PORT
http://172.18.0.2:30624
```

```bash
# check the APISIX containerId
root@kind-control-plane:/# crictl ps
CONTAINER           IMAGE               CREATED             STATE               NAME                      ATTEMPT             POD ID              POD
711b3c8195019       7d2986d153861       3 minutes ago       Running             apisix                    0                   78ea3ecfb1fc2       apisix-fbc449d4-bs8tz

# attach to APISIX container
root@kind-control-plane:/# crictl exec -it 711b3c8195019 bash

# create a router
bash-5.1# curl http://127.0.0.1:9180/apisix/admin/routes/1  \
-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "methods": ["GET"],
    "uri": "/anything",
    "plugins": {},
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "172.18.0.1:80": 1
        }
    }
}'
{"node":{"key":"\/apisix\/routes\/1","value":{"upstream":{"hash_on":"vars","pass_host":"pass","type":"roundrobin","nodes":{"172.18.0.1:80":1},"scheme":"http"},"plugins":{},"create_time":1689133203,"id":"1","status":1,"priority":0,"methods":["GET"],"update_time":1689153936,"uri":"\/anything"}},"action":"set"}
```

```bash
$ wrk -t30 -c100 -R200 -d30s http://172.18.0.2:30624/anything  
```

```bash
root@kind-control-plane:/# crictl exec -it 711b3c8195019 bash
bash-5.1# cat test_error.log |grep semaphore
```

**Comment by Sn0rt** [comment]
It seems to be a problem with the DNS client, not a user plugin.  @kingluo  @monkeyDluffy6017  Do you have any suggestion?

**Comment by wklken** [comment]
> 2023/07/20 06:39:02 [error] 32#32: in lua semaphore gc wait queue is not empty while the semaphore 00007F3B87D78598 is being destroyed

apisix: 3.2.0

when we sync settings into apisix etcd in parallel(bulk put routers and services), few error logs showed.
(only when syncing settings)

**Comment by yingtao-martin** [comment]
10.60.13.51:30050 "GET /spl/mes/base/mgt/staff/selector?t=1713272826859&orgCode=G02&carrierCode=C0028&vehicleId=1778468996493234178 HTTP/1.1" 200 4825 0.001 "http://10.60.13.51:30001/tms/index.html" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.5735.289 Safari/537.36" 192.168.0.173:9080 200 0.001 "http://10.60.13.51:30050"
[error] 44#44: *11294423 [lua] client.lua:116: reportServiceInstance(): Instance report fails, connection refused, context: ngx.timer
[error] 44#44: *11300312 [lua] init.lua:290: get_url:ns/instance/list?healthyOnly=true&serviceName=xxx-coop-cus&namespaceId=5cab192c-cda1-43cd-a6cc-d767c34e3d0d&groupName= err:failed to parse domain: failed to parse domain, context: ngx.timer
10.60.13.51:30050 "OPTIONS /spl/mes/disp/mgt/dispatch/getStatusCount HTTP/1.1" 200 5 0.000 "http://10.60.13.55:30001/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36 Edg/123.0.0.0" - - - "http://10.60.13.51:30050"
10.60.13.51:30050 "POST /spl/mes/disp/mgt/dispatch/getStatusCount HTTP/1.1" 200 2780 0.004 "http://10.60.13.55:30001/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36 Edg/123.0.0.0" 192.168.0.170:9080 200 0.004 "http://10.60.13.51:30050"
10.60.13.51:30050 "OPTIONS /spl/mes/disp/mgt/dispatch/dispatchList HTTP/1.1" 200 5 0.000 "http://10.60.13.55:30001/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36 Edg/123.0.0.0" - - - "http://10.60.13.51:30050"
10.60.13.51:30050 "POST /spl/mes/disp/mgt/dispatch/dispatchList HTTP/1.1" 200 8773 0.045 "http://10.60.13.55:30001/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36 Edg/123.0.0.0" 192.168.0.164:9080 200 0.045 "http://10.60.13.51:30050"


Environment
APISIX version (run apisix version): 2.13.0
Operating system (run uname -a): Linux apisix-7dc879cc4-hwm7g 3.10.0-1160.15.2.el7.x86_64
OpenResty / Nginx version (run openresty -V or nginx -V): openresty/1.21.4.1
etcd version, if relevant (run curl http://127.0.0.1:9090/v1/server_info): 3.5.0
APISIX Dashboard version, if relevant:
Plugin runner version, for issues related to plugin runners: no
LuaRocks version, for installation issues (run luarocks --version):

**Comment by virgo915** [comment]
This issue has also occurred in my company, with the same phenomenon as the original poster: some requests succeed, while others time out with a 499 response code. We reviewed a portion of the logs from the time the previous failure occurred and found a persistent request with a very large response body (500KB). Because it was a persistent connection, the client initiating this request kept sending the same request to this process, causing it to freeze.

**Comment by Baoyuantop** [comment]
Hi @virgo915, the original issue was created too long ago. Please create a new issue to describe your problem.
